### PR TITLE
Generate full debug info

### PR DESCRIPTION
**Pull request by umanwizard**
Recently, @petrosagg and @antiguru tried to debug a release blocker in staging using a core dump, and found themselves unable to do so, because the core dump didn't contain information about local variables. This PR should mitigate that issue.

To check the impact on CI times, I experimented setting `RUSTFLAGS="-C debuginfo=2"` and comparing with debuginfo=1.

Here are the results on my home desktop:

| Measurement | debuginfo=2 | debuginfo=1 |
|--------|--------|--------|
| `rm -rf target && time cargo build --release` | 13:39.95 | 10:24.29 |
| `du -hs target/release/clusterd` | 3.6G | 2.0G |
| `du -hs target/release/environmentd` | 2.3G | 1.3G |
| `du -hs target` | 27G | 18G |

In conclusion, the build is a few minutes slower, and we produce a bit bigger binaries. However, I think the improved ability to debug issues in prod or staging is worth the cost.

### CODE REVIEW

**Comment by benesch** [comment]
They’re not a bit bigger: they’re nearly twice the size! I don’t think we should compromise the common case for the exception case to this degree. If we need this, I think it is time to set up split debuginfo and a debuginfod server, or whatever the state of the art for these things is today.

File: Cargo.toml [hunk 1]
```diff
@@ -99,13 +99,7 @@ object = { opt-level = 3 }
 rustc-demangle = { opt-level = 3 }
 
 [profile.release]
-# Emit only the line info tables, not full debug info, in release builds, to
-# substantially reduce the size of the debug info. Line info tables are enough
-# to correctly symbolicate a backtrace, but do not produce an ideal interactive
-# debugger experience. This seems to be the right tradeoff for release builds:
-# it's unlikely we're going to get interactive access to a debugger in
-# production installations, but we still want useful crash reports.
```
Comment (benesch) [inline Cargo.toml:107]: If we do decide to go forward with this, the comment needs to be updated with the new justification, not removed.

**Comment by petrosagg** [comment]
Incidentally I also tried compiling in release mode with debug symbols and
the actual binaries were much smaller (for both debug 1 and debug 2 modes).
I built using `mzcompose run default` on some random test. I won't be near
my computer today but I can report my numbers tomorrow (unless someone does
the mzcompose builds)

**Comment by umanwizard** [comment]
@benesch what is the actual negative impact of larger binaries, so that we can evaluate the tradeoffs quantitatively? Is it increased CI time due to Docker slowness, runtime slowdown, or something else?

I agree that it would be ideal if we implemented something based on symbol splitting and debuginfod, but as far as I know, nobody is working on that.

**Comment by petrosagg** [comment]
These are the size differences of the binaries in the docker image that I observed:

**From a local docker image built with mzcompose and with `debug = 2`:**

```
materialize@29a3fa4d75d6:/$ ls -lah /usr/local/bin/*
-rwxr-xr-x 1 root root 598M Jun  9 11:23 /usr/local/bin/clusterd
-rwxr-xr-x 1 root root 402M May 16 02:18 /usr/local/bin/cockroach
-rwxr-xr-x 1 root root 3.0K Apr 17 13:41 /usr/local/bin/entrypoint.sh
-rwxr-xr-x 1 root root 573M Jun  9 11:23 /usr/local/bin/environmentd
```

**From the canonical `materialize/materialized:0.57.0` image:**

```
[petrosagg@doggo:~]$ docker run -it --entrypoint=/bin/bash 0cf67d324134 -c /bin/bash
materialize@64e942cbee0d:/$ ls -lah /usr/local/bin/*
-rwxr-xr-x 1 root root 423M Jun  5 21:15 /usr/local/bin/clusterd
-rwxr-xr-x 1 root root 402M May 16 02:18 /usr/local/bin/cockroach
-rwxr-xr-x 1 root root 3.0K Jun  5 21:02 /usr/local/bin/entrypoint.sh
-rwxr-xr-x 1 root root 397M Jun  5 21:15 /usr/local/bin/environmentd
```

**Comment by benesch** [comment]
> @benesch what is the actual negative impact of larger binaries, so that we can evaluate the tradeoffs quantitatively? Is it increased CI time due to Docker slowness, runtime slowdown, or something else?

These are the issues off the top of my head:

* Longer compile times (both in CI and locally when building in release mode)
* Longer CI times in subsequent jobs (since pulling images takes longer)
* Longer iteration cycles when using `mzcompose` locally (since pulling images takes longer)
* Slows down any Docker-based demos (e.g. Instruqt) that need to pull the image before the demo can start

**Comment by necaris** [comment]
Is it prohibitively complex to build a debuginfo artifact with a separate tag, so we can opt into them (e.g. in our staging env)?

**Comment by benesch** [comment]
> Is it prohibitively complex to build a debuginfo artifact with a separate tag, so we can opt into them (e.g. in our staging env)?

No, but that has other downsides: namely that staging is not a perfect copy of prod!

@umanwizard and I chatted on Slack and he's going to rig up our CI to publish debuginfos to an S3 bucket, where Sentry and GDB can automatically download them.
